### DISCUSSION TITLE
In what scenario would stdin come from a socket?

### DISCUSSION THREAD

**Discussion by smahm006** [BurntSushi/ripgrep#2448] [Q&A] [answered]
Hello everyone,

I am a Python programmer trying to learn Rust. I like to learn by studying well made projects and for Rust I am using the ripgrep project to learn the language (Thank you BurntSushi for leaving this open source!)

I started with the grep-cli crate and just from digging into it's functions I realize there are a lot of low level concepts that I need to understand. The first function I started with is the "pub fn is_readable_stdin()" (in crates/cli/src/lib.rs) and that took me on a whole journey of a **basic** understanding what file descriptors are, how inter-process communications work and how everything is a file in Unix. 

For now I understand why piping ( cat test.txt | rg "test file") and stdin redirection (rg "test file" < test.txt) make is_stdin_readable() true. 

But there is also a check to see if stdin is coming from a socket (ft.is_socket()) but I am  getting stuck trying to find an example where that would be applicable.

From https://unix.stackexchange.com/questions/16311/what-is-a-socket: 

> A socket is a pseudo-file that represents a network connection. Once a socket has been created (identifying the other host and port), writes to that socket are turned into network packets that get sent out, and data received from the network can be read from the socket.

And there are examples of opening a socket `nc -lU socket.sock` and writing to it `socat -T 3 unix-connect:socket.sock 2` but I cannot find how to read from a socket, hence the question of the title.

**Comment by BurntSushi** [comment, marked as answer]
This is the commit that added the `ft.is_socket()` predicate: https://github.com/BurntSushi/ripgrep/commit/020c5453a5880257c87949030550d24c596f5c8d

The commit message also basically answers your question. It sounds like PowerShell uses a socket instead of a fifo. I do not know why it's implemented that way.

See also https://github.com/BurntSushi/ripgrep/issues/1741, which is where the bug was originally reported.

**Reply by smahm006** [reply]
Ah makes sense, so generally stdin does not come from sockets.

Thanks for the clarification!

**Reply by BurntSushi** [reply]
Yeah it's definitely a strange case.
